smolyay/samples.py

```diff
@@ -508,3 +511,580 @@ def _create(self):
+    """Points generated using all combinations of unidimensional points.
+
+    Depending on the dimensionality, points and provided by
+    :class:`UnidimensionalPointSet`, make full tensor grids.
+    :meth:`generates_points` generates a full tensor grid
+
+    Parameters
+    ----------
+    unique_points : list of :class:UnidimensionalPointSet
+        unique set of 1D points
+    """
+
+    def _create(self):
+        """Generating the tensor grid
+
+        Using the parameters of the class, generate a grid set using all
+        combinations of unidimensional points
+        """
+        num_points = numpy.prod([len(p) for p in self._point_sets])
+        self._points = numpy.zeros((num_points, len(self._point_sets)), dtype=float)
+        for i, point in enumerate(itertools.product(*self._point_sets)):
+            self._points[i] = point
+
+
+class SmolyakSparseProductPointSet(PointSetProduct):
+    """Points generated from sparse combinations of unidimensional points.
+
+    Depending on the dimensionality, and levels in each point set in
+    param:`point_sets`, makes sparse smolyak grids.
+
+    Parameters
+    ----------
+    unique_points : list of :class:UnidimensionalPointSet
+        unique set of 1D points
+    """
+
+    def _create(self):
+        """Generating the sparse smolyak grid
+
+        Using the parameters of the class, generate a grid set using all
+        valid smolyak combinations of unidimensional points
+        """
+        # reset points
+        self._points = None
+
+        # find number of levels per dimension
+        num_levels_per_dim = [ob.num_levels for ob in self._point_sets]
+        max_num_levels = max(num_levels_per_dim)
+
+        # get the combinations of levels based on maximum possible number of levels
+        level_combinations = []
+        for sum_of_levels in range(max_num_levels):
+            level_combinations.extend(
+                list(
+                    _generate_compositions(
+                        sum_of_levels, self.num_dimensions, include_zero=True
+                    )
+                )
+            )
+        level_combinations = numpy.array(level_combinations)
+
+        # remove combinations where a dimension exceeds its number of levels
+        # only check if point sets have different numbers of levels
+        if min(num_levels_per_dim) != max_num_levels:
+            valid_comb = numpy.all(numpy.less(level_combinations, num_levels_per_dim), axis=1)
+            level_combinations = level_combinations[valid_comb]
+
+        # generate sets of points based on combinations of levels
+        for level_comb in level_combinations:
+            level_point_combinations = [
+                self._point_sets[d].level(level) for d, level in enumerate(level_comb)
+            ]
+            num_points = numpy.prod([len(p) for p in level_point_combinations])
+            points_ = numpy.zeros((num_points, len(level_point_combinations)), dtype=float)
```

Comment:
Can you use `self.num_dimensions` here?
```suggestion
            points_ = numpy.zeros((num_points, self.num_dimensions), dtype=float)
```

Reply:
yes